### ISSUE TITLE
VCF output for PE data is wonky

### ISSUE THREAD

**Issue by dereneaton** [dereneaton/ipyrad#225]
Reported on gitter.im

denovo
```
locus_3925 142 . G A,T 13 PASS NS=4;DP=33 GT:DP:CATG 0/2:9:1,1,3,4 0/1:7:7,0,0,0 0/2:9:0,1,3,5 ./.:0:0,0,0,0 ./.:0:0,0,0,0 ./.:0:0,0,0,0 ./.:0:0,0,0,0 ./.:0:0,0,0,0 ./.:0:0,0,0,0 ./.:0:0,0,0,0 ./.:0:0,0,0,0 3/1:8:3,4,0,1
```
I also get malformed CATG fields (missing count): locus_17982 155 . A C 13 PASS NS=12;DP=12794 
```
GT:DP:CATG 1/0:946:154,791,1,0 0/0:1091:141,950,0,0 0/0:1127:152,974,1,0 0/0:1114:138,975,1,0 1/0:1110:761,348,0,1 0/0:841:0,0,1,840 1/0:679:475,203,1,0 1/0:977:664,312,1,0 0/0:1367:0,0,3,1364 0/0:1055:0,0,0,1055 0/0:1264:140,1123,1, 0/0:1223:160,1060,3,
```

**Comment by ODiogoSilva** [comment]
I am also having a similar issue but for SE RAD data. For some reason, I have heterozygous genotypes in my VCF with base counts for only one nucleotide:

`0/1:34:0,34,0,0`

The proportion of these weird genotypes is rather small on this particular VCF, but I don't know if this site should be considered homozygous or heterozygous.

**Comment by StuntsPT** [comment]
I just re-ran using `ipyrad` 0.6.9 and the issue remains...

**Comment by isaacovercast** [comment]
I have not been able to reproduce this for either SE RAD or PE ddRAD on real data. @StuntsPT Can you dropbox me the hdf5 file? Or better a chunk of the raw data and the params file you used.

**Comment by StuntsPT** [comment]
@isaacovercast I sent you a link via gitter private chat.

**Comment by isaacovercast** [comment]
__tldr;__: Yes the genotype counts are wrong in the vcf file, but the genotypes themselves are correct. Heterozygous sites ARE correct even if the counts are wrong.

Got good news and bad news. The good news is the consens.hdf5 file is intact and looking good. The other good news is this means that all vcf files we've been creating up to now are still ok, in terms of the actual genotypes. The bad news is the output of make_vcf must be fscking something and I'm not super familiar with that code so it might take some time to fix. Still not sure exactly what's happening or why his is doing this, but here's the output for this snp for two individuals, from the vcf output above the "-11" individual has the good catg info and the "16" individual has the homozygous looking catg:
```
>>> io5["catgs"][37200][-11][34]
array([106,   0,  50,   1], dtype=uint32)
>>> io5["catgs"][37200][16][34]
array([47,  0, 19,  0], dtype=uint32)
>>> io5["seqs"][37200][16][34]
'Y'
>>> io5["seqs"][37200][-11][34]
'Y'
```

**Comment by isaacovercast** [comment]
The "bad" individual has the incorrect genotype counts (" [66,  0,  0,  0]") in the neighborhood of the bad base:
```
>>> io5["catgs"][37200][16][30:40]
array([[ 0, 66,  0,  0],
       [ 0,  0, 66,  0],
       [ 0,  0,  0, 66],
       [ 0,  0,  0, 66],
       [47,  0, 19,  0],
       [ 0,  0,  1, 65],
       [66,  0,  0,  0],
       [ 0,  0,  0, 66],
       [ 0,  0,  0, 66],
       [ 0,  0, 66,  0]], dtype=uint32)
```

**Comment by StuntsPT** [comment]
The .loci writer may be acting up too, since the issue is also revealed there...

**Comment by eaton-lab** [comment]
I haven't had a chance to look at this yet, but one idea is that the problem may be related to indels, which are imputed into the catg array after step6 aligning. If an indel is not properly imputed then the base counts would not line up properly with the base calls.

**Comment by isaacovercast** [comment]
Also __NB:__ the actual cluster number for this locus inside the clust.hdf5 file is __37200__, and not `18106` as reported in the .loci file, this took me a while to figure out. 32700 is a conspicuously round number, perhaps an `optim` edge case? Though my instinct was something about indels as well.

**Comment by StuntsPT** [comment]
That's great news. It means I can carry on with my analyses! The read counts are not vital, I was just worried that the genotypes could be incorrectly called.

**Comment by StuntsPT** [comment]
Just for the record, I have provided via private gitter.im message the required sample files to reproduce the issue.

**Comment by isaacovercast** [comment]
omg how is this still an issue? I had hoped the update to v0.9 had fixed this, but apparently not.

This was reported on gitter and I got the user to share data necessary to reproduce:

`RAD_0   6       loc0_pos5       C       T       13      PASS    NS=114;DP=3165  GT:DP:CATG      0/0:20:20,0,0,0 0/0:44:0,0,44,0 0/0:16:16,0,0,0 0/0:25:0,0,25,0 0/0:21:0,0,21,0       0/0:15:15,0,0,0 0/0:8:0,0,8,0`

**Comment by isaacovercast** [comment]
There appears to be no rhyme or reason to the depth issue. It's not a sample issue because I looked at 2 different samples and in both of them sometimes the `catg`s line up with the sequence and sometimes they don't. It's not locus specific because within any given locus I observe samples with both good and bad catg data. It appears unrelated to indels, because I can see sometimes a sample at a locus with tons of indels still has correct catg info, and sometimes a sample at a locus with complete sequence (except 2 indels at the end) has bad catg info.

**Comment by isaacovercast** [comment]
Simulated data appears unaffected (naturally).

**Comment by isaacovercast** [comment]
So, in consens_se.py the catgs are calculated from `self.arrayed` on line 709. The sequence that gets written out comes from `self.consens`. `self.arrayed` is created inside `build_consens_and_array` and then `self.consens` is generated from arrayed by `base_caller`, then a bunch of other stuff happens, and somewhere in there `arrayed` and `consens` diverge. Here are all the places arrayed gets manipulated:

Line 611:
```
            self.consens = self.consens[ltrim:rtrim + 1]
            self.arrayed = self.arrayed[:, ltrim:rtrim + 1]
```
Line 637 (inside `mask_repeats`):
```
        # apply filter
        self.consens = self.consens[keep]
        self.arrayed = self.arrayed[:, keep]
```
This all looks legit.

`storealleles` manipulates `consens`, but I can see that this problem crops up both for sequences with 1 and with 2 alleles, so it can't be this.

**Comment by isaacovercast** [comment]
FUCK! I cracked it. It's GBS.

Here is a bad base:
```
RAD_9   8       loc9_pos7       G       T       13      PASS    NS=2;DP=26      GT:DP:CATG      1/1:9:0,0,9,0   0/0:17:0,0,17,0 ./.:0:0,0,0,0   ./.:0:0,0,0,0
```
Here is the locus in the clust_database.fa:
```
>S197batch2_10079
TAATATTTCTTATAGTTGTGTCAGATTAGAACAATCTGCTTGTTGTTTGGTGTAGTTATGACTTCAGAAACATAGAAGAAGACAGGCAGTTGCATTTGTTTTGTCAGCAATTGTCTGGCAGTTTCTAGTTGGACAGAGACTGATnnnnGCAGTTTCTAGTTGGACAGAGACTGATTAGAAGAATCTGCTTGTTGTTTGGTATAGTTATAACTTCAGAAACTAAGAAGAAGATAGGCAGTTGCATTTGTTTTGTCAGCCATTTTCTGGCAGTTTGTAGCTGGACGGAGACTTA
>S226zAbatch2_6418
TAATATTTCTGATAGTTGTGTCAGATTAGAACAATCTGCTTGTTGTTTGGCGTAGTTATGACTTCAGAAACATAGAAGAAGACAGGCAGTTGCATTTGTTTTGTCAGCAATTGTCTGGCAGTTTCTAGTTGGACAGAGACAGATnnnnGCAGTTTCTAGTTGGACAGAGACAGATTAGAAGAATCTGCTTGTTGTTTGGTATAGTTATAACTTCAGAAACTAAGAAGAAGATAGGCAGTTGCATTTGTTTTGTCAGCCATTTTCTGGCAGTTTGTAGCTGGACGGAGACTTA
```
Here are the first 17 bases from the catg for S197batch2 (matches the proximal end of the sequence in the clust db):
```
[[0 0 9 0]
 [0 9 0 0]
 [0 9 0 0]
 [0 0 9 0]
 [0 9 0 0]
 [0 0 9 0]
 [0 0 9 0]
 [0 0 9 0]
 [9 0 0 0]
 [0 0 8 1]
 [0 0 9 0]
 [0 9 0 0]
 [0 0 9 0]
 [0 9 0 0]
 [0 0 0 9]
 [0 0 9 0]
 [0 0 9 0]]
```
Here are the first 17 bases from the catg for S226zAbatch2 (revcomp matches the last 17 bases of the distal end of the sequence in the clust db):
```
[[ 0  0 17  0]
 [ 0 17  0  0]
 [ 0 17  0  0]
 [ 0  0  0 17]
 [ 0  0 17  0]
 [17  0  0  0]
 [ 0  0 17  0]
 [17  0  0  0]
 [17  0  0  0]
 [ 0  0  0 17]
 [ 0  0 17  0]
 [17  0  0  0]
 [17  0  0  0]
 [ 0 17  0  0]
 [ 0  0  0 17]
 [17  0  0  0]
 [ 0  0 17  0]]
```
Still don't know how to fix it, but damn, i'm super happy to have figured it out and it seems like it might be an easy fix.....

**Comment by isaacovercast** [comment]
SHIIIIIIIIITTTTTTTTTTTTTTTTTTTTTTTTT! It's not easy to fix. In clustmap_across.py GBS data gets clustered on both strands, so sequences can 'flip' to match the strand of the seed, which obviously flips the orientation of the sequence and breaks the connection with the catg info constructed during step 5. Damn.

**Comment by isaacovercast** [comment]
The fix for this will be to parse the utemp file during step 6 and step through each catg file for each sample and reverse the depth info for hits that match on the negative strand. This only needs to be done for GBS data, and probably should happen at the end of step 6. This will be super annoying, and a lot of work, so I will file this under "will fix when someone pays me to do it." ;p
